### PR TITLE
[WIP] Proposal: Move docker-registry to separate repository

### PR DESCRIPTION
**Pull request by legionus**

### CODE REVIEW

**Comment by legionus** [comment]
@mfojtik @miminar @soltysh PTAL

File: docs/proposals/docker-registry-split.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+### etcd
+
+The docker-registry does not have its own `etcd`. All metadata are stored to the `origin` database using
+client API.
+
+### code-base
+
+The `origin` interacts with the docker-registry only through client library (no internal calls),
+but this cannot be said about docker-regisrty.
```
Comment (miminar) [inline docs/proposals/docker-registry-split.md]: `s/regisrty/registry/`

**Comment by miminar** [comment]
- Pick-up a new registry repository name.
- How will this affect spec file?
- What will we do with registry-related end-to-end tests?
  - Do we want to turn them into extended tests before the split happens? See related issue https://github.com/openshift/origin/issues/10813

**Comment by legionus** [comment]
> - Pick-up a new registry repository name.

origin-registry ?

> - How will this affect spec file?

probably we will have own spec-file in the new repository.

> - What will we do with registry-related end-to-end tests?

I think they should stay in origin.

>  - Do we want to turn them into extended tests before the split happens? See related issue https://github.com/openshift/origin/issues/10813

I don't know yet.

File: docs/proposals/docker-registry-split.md (hunk 1 above)
Comment (legionus) [inline docs/proposals/docker-registry-split.md]: fixed

**Comment by mfojtik** [comment]
@smarterclayton @ncdc PTAL

File: docs/proposals/docker-registry-split.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+## Constraints and Assumptions
+
+The docker-registry is highly integrated in a common code-base so it is very difficult to describe
+all the difficulties of separation.
+
+## Investigation
+
+### etcd
+
+The docker-registry does not have its own `etcd`. All metadata are stored to the `origin` database using
```
Comment (mfojtik) [inline docs/proposals/docker-registry-split.md]: worth to mention the problem with storage if we continue using the openshift etcd.

File: docs/proposals/docker-registry-split.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+client API.
+
+### code-base
+
+The `origin` interacts with the docker-registry only through client library (no internal calls),
+but this cannot be said about docker-registry.
+
+1. docker-registry should use [clientsets](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/generating-clientset.md)
+of `origin`.
+2. docker-registry should use [clientsets](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/generating-clientset.md)
```
Comment (mfojtik) [inline docs/proposals/docker-registry-split.md]: @deads2k

File: docs/proposals/docker-registry-split.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+of `kubernetes`.
+3. We need to create versioned client (clientsets?) for docker-registry as well.
```
Comment (mfojtik) [inline docs/proposals/docker-registry-split.md]: is `docker/engine-api` not enough? or we are talking about adding more API endpoints to registry?

File: docs/proposals/docker-registry-split.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+4. We need to split some utilities to reduce the number of dependencies to `origin`.
```
Comment (mfojtik) [inline docs/proposals/docker-registry-split.md]: are you talking about the build utilities?

File: docs/proposals/docker-registry-split.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+5. We need to make common framework for tests. Currently docker-registry uses a common test utilities and clients.
```
Comment (mfojtik) [inline docs/proposals/docker-registry-split.md]: we should also start enabling the upstream docker-registry tests. we also need to enable the tests for different storage drivers to prove the storage works (remember GCS? :-)

i think this should be also mentioned here as part of the new infrastructure.

File: docs/proposals/docker-registry-split.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+### testing
+
+Use versioning clients will allow us maintain compatibility between repositories. In this case,
+in the jenkins job for `origin` we can build the image from the separate repository before build and
+test `origin` itself. Right now we make docker-registry image from the `origin` repository.
```
Comment (mfojtik) [inline docs/proposals/docker-registry-split.md]: this should also reduce the size of docker-registry right? (when we start not importing stuff from origin).

do we also need a regression tests (iow. registry vs. old version of origin).

**Comment by legionus** [comment]
[(8) [registry] R&D: Registry repository split](https://trello.com/c/rRQPZk9v/789-8-registry-r-d-registry-repository-split)

**Comment by pweil-** [comment]
closing due to inactivity

**Pull request closed without merging**